hummingbot/connector/exchange/xrpl/xrpl_api_order_book_data_source.py

```diff
@@ -0,0 +1,236 @@
+import asyncio
+import time
+from decimal import Decimal
+
+# import time
+from typing import TYPE_CHECKING, Any, Dict, List, Optional
+
+# XRPL imports
+from xrpl.asyncio.clients import AsyncWebsocketClient
+from xrpl.models.requests import BookOffers, Subscribe, SubscribeBook
+from xrpl.utils import get_order_book_changes, ripple_time_to_posix
+
+from hummingbot.connector.exchange.xrpl import xrpl_constants as CONSTANTS
+from hummingbot.connector.exchange.xrpl.xrpl_order_book import XRPLOrderBook
+from hummingbot.core.data_type.common import TradeType
+from hummingbot.core.data_type.order_book_message import OrderBookMessage
+from hummingbot.core.data_type.order_book_tracker_data_source import OrderBookTrackerDataSource
+from hummingbot.core.utils.async_utils import safe_gather
+from hummingbot.core.web_assistant.web_assistants_factory import WebAssistantsFactory
+from hummingbot.logger import HummingbotLogger
+
+if TYPE_CHECKING:
+    from hummingbot.connector.exchange.xrpl.xrpl_exchange import XrplExchange
+
+
+class XRPLAPIOrderBookDataSource(OrderBookTrackerDataSource):
+    _logger: Optional[HummingbotLogger] = None
+
+    def __init__(self,
+                 trading_pairs: List[str],
+                 connector: 'XrplExchange',
+                 api_factory: WebAssistantsFactory):
+        super().__init__(trading_pairs)
+        self._connector = connector
+        self._api_factory = api_factory
+        self._trade_messages_queue_key = CONSTANTS.TRADE_EVENT_TYPE
+        self._diff_messages_queue_key = CONSTANTS.DIFF_EVENT_TYPE
+        self._snapshot_messages_queue_key = CONSTANTS.SNAPSHOT_EVENT_TYPE
+        self._xrpl_client = AsyncWebsocketClient(self._connector.node_url)
+        self._open_client_lock = asyncio.Lock()
+
+    async def get_last_traded_prices(self,
+                                     trading_pairs: List[str],
+                                     domain: Optional[str] = None) -> Dict[str, float]:
+        return await self._connector.get_last_traded_prices(trading_pairs=trading_pairs)
+
+    async def _request_order_book_snapshot(self, trading_pair: str) -> Dict[str, Any]:
+        """
+        Retrieves a copy of the full order book from the exchange, for a particular trading pair.
+
+        :param trading_pair: the trading pair for which the order book will be retrieved
+
+        :return: the response from the exchange (JSON dictionary)
+        """
+        # Create a client to connect to the test network
+        # client = self._client
+        base_currency, quote_currency = self._connector.get_currencies_from_trading_pair(trading_pair)
+
+        async with self._open_client_lock:
+            try:
```

Comment:
It would be better to remove this try / catch so that you don't end up with an empty order book when one snapshot fails. We already catch the exception in ```listen_for_order_book_snapshots```